clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -513,10 +518,25 @@ ProducerIdAndEpoch producerIdAndEpoch() {
     }
 
     boolean producerIdOrEpochNotMatch(ProducerBatch batch) {
-        ProducerIdAndEpoch idAndEpoch = this.producerIdAndEpoch;
+        // Verifies that the producer id/epoch of the batch matches the current one
+        // of the partition.
+        ProducerIdAndEpoch idAndEpoch = topicPartitionBookkeeper.getPartition(batch.topicPartition).producerIdAndEpoch;
```

Comment:
It looks like `producerIdOrEpochNotMatch` has one caller in `handleFailedBatch`. Now that we are keeping the producerId/epoch consistent with the inflight batches, I guess this check is more of a state validation than an essential check. I'm wondering if we can either remove the check and rely on our internal invariant or change it from logging a debug message to raising an IllegalState error.

Reply:
Good point. I think that we can just remove it. It does not seem to bring much.